net/sync_dag.go

```diff
@@ -58,37 +48,48 @@ func syncDAG(ctx context.Context, bserv blockservice.BlockService, block *corebl
 		return err
 	}
 
-	ssb := builder.NewSelectorSpecBuilder(basicnode.Prototype.Any)
-	matchAllSelector, err := ssb.ExploreRecursive(selector.RecursionLimitNone(), ssb.ExploreUnion(
-		ssb.Matcher(),
-		ssb.ExploreAll(ssb.ExploreRecursiveEdge()),
-	)).Selector()
+	err = loadBlockLinks(ctx, lsys, block)
 	if err != nil {
 		return err
 	}
+	return nil
+}
 
-	// prototypeChooser returns the node prototype to use when traversing
-	prototypeChooser := func(lnk ipld.Link, lnkCtx ipld.LinkContext) (ipld.NodePrototype, error) {
-		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
-			return tlnkNd.LinkTargetNodePrototype(), nil
-		}
-		return basicnode.Prototype.Any, nil
-	}
-	// preloader is used to asynchronously load blocks before traversing
-	//
-	// any errors encountered during preload are ignored
-	preloader := func(pctx preload.PreloadContext, l preload.Link) {
-		go lsys.Load(linking.LinkContext{Ctx: pctx.Ctx}, l.Link, coreblock.SchemaPrototype) //nolint:errcheck
-	}
-	config := traversal.Config{
-		Ctx:                            ctx,
-		LinkSystem:                     lsys,
-		LinkVisitOnlyOnce:              true,
-		LinkTargetNodePrototypeChooser: prototypeChooser,
-		Preloader:                      preloader,
-	}
-	visit := func(p traversal.Progress, n datamodel.Node) error {
-		return nil
+// loadBlockLinks loads the links of a block recursively.
+//
+// If it encounters errors in the concurrent loading of links, it will return
+// the first error it encountered.
+func loadBlockLinks(ctx context.Context, lsys linking.LinkSystem, block *coreblock.Block) error {
+	ctx, cancel := context.WithTimeout(ctx, syncDAGTimeout)
+	defer cancel()
+
+	var asyncErr error
+	var asyncErrOnce sync.Once
+	wg := &sync.WaitGroup{}
+	for _, lnk := range block.Links {
+		wg.Add(1)
+		go func(lnk coreblock.DAGLink) {
+			defer wg.Done()
+			nd, err := lsys.Load(linking.LinkContext{Ctx: ctx}, lnk, coreblock.SchemaPrototype)
+			if err != nil {
+				asyncErrOnce.Do(func() {
+					asyncErr = err
+				})
+				cancel()
+				return
+			}
+			linkBlock, err := coreblock.GetFromNode(nd)
+			if err != nil {
+				asyncErrOnce.Do(func() {
+					asyncErr = err
+				})
+				cancel()
```

Comment:
question: `cancel()` is only for err cases and not to be run using `defer`?

Reply:
It is used with `defer` but also explicitly called on error to help stop the goroutines.

Reply:
Ah makes sense. I missed the defer above.